torch/utils/_as_strided.py

```diff
@@ -0,0 +1,492 @@
+
+First, take the source tensor and put it into canonical form by:
+
+1. Removing all size-1 and stride-0 dimensions: For each stride-0 dimension,
+   narrow it to size 1, then squeeze all size-1 dimensions at once.
+2. Permuting the tensor so the strides are in strictly descending order.
+   This is always possible as all size-1 dimensions are removed and we
+   required the input to be non-overlapping.
+3. Flattening all contiguous dims, so that every pair of adjacent
+   dims is non-contiguous with each other (cannot be flattened).
+
+For any non size-1/stride-0 dimension in the target tensor, we can uniquely
+assign it to the only canonical source dim that could have created it.
+Specifically, for any dimension (x,):(a,), any target stride s such that
+a <= s <= (x-1) * a could only have been generated from this source dim.
+(x*a is excluded as it can only be generated by a size-1 dim!  And it is
+important to be precise on the bound here because it is not
+guaranteed that on a canonical source tensor (x,y):(a,b) that y*b <= a).
+So our plan is to process each canonical source dim one-by-one and translate
+them into the target dimensions.
+
+For each canonical dimension, we may have multiple target dimensions that map
+to it. We process these target dimensions left to right (largest stride first).
+
+For each target dimension (except the last one from this canonical dim):
+  1. We unflatten to create the target dimension and a "remainder" dimension.
+     The key insight: unflatten(x*y, (x, y)) gives strides (y*a, a) where a
+     is the canonical stride. To get the target stride s for the first dim,
+     we need y*a = s, so we choose y = s/a.
+  2. This creates dimension with the correct stride. We then move to the next
+     target dimension, which will be carved out of the "remainder" dimension.
+  3. If the needed size (target_size * y) exceeds available size, we return
+     NotImplemented.
+
+For the last target dimension from this canonical dim:
+  1. Its stride must match the canonical stride, otherwise we return NotImplemented.
+  2. Narrow to the final target size.
+
+Limitations:
+  - Slice operations with nontrivial step (e.g., [::2]) are not currently
+    supported. This means some valid view sequences cannot be reconstructed.
+"""
+
+import functools
+import types
+
+import torch
+
+
+def _canonicalize_tensor(tensor: torch.Tensor) -> torch.Tensor:
+    """
+    Canonicalize a tensor by:
+    1. Narrowing stride-0 dimensions to size 1
+    2. Squeezing all size-1 dimensions
+    3. Permuting so strides are in descending order
+    4. Flattening contiguous dimensions
+
+    Returns a tensor with strictly descending strides where no two adjacent
+    dims are contiguous with each other, and every dim has size > 1.
+    """
+    result = tensor
+
+    # Narrow stride-0 dimensions to size 1
+    for dim in reversed(range(result.ndim)):
+        if result.stride(dim) == 0:
+            result = result.narrow(dim, 0, 1)
```

Comment:
nit: in case it is helpful, we can use select() here to narrow+squeeze in a single op.